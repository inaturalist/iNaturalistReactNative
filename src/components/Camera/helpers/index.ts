import { Alert } from "react-native";
import { CameraRuntimeError } from "react-native-vision-camera";

import { log } from "../../../../react-native-logs.config";

const logger = log.extend( "CameraView" );

type Event = {
  log: string;
}

const handleClassifierError = ( error: CameraRuntimeError ) => {
  // When we hit this error, there is an error with the classifier.
  logger.error( "handleClassifierError: ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleDeviceNotSupported = ( error: CameraRuntimeError ) => {
  // When we hit this error, something with the current device is not supported.
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleCaptureError = ( error: CameraRuntimeError ) => {
  // When we hit this error, taking a photo did not work correctly
  logger.error( "handleCaptureError: ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleCameraError = ( error: CameraRuntimeError ) => {
  // This error is thrown when it does not fit in any of the above categories.
  logger.error( "handleCameraError ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleLog = ( event: Event ) => {
  logger.debug( event.log );
};

export {
  handleCameraError,
  handleCaptureError,
  handleClassifierError,
  handleDeviceNotSupported,
  handleLog
};
