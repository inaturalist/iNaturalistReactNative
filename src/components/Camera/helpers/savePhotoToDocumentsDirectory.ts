import {
  rotatedOriginalPhotosPath
} from "appConstants/paths";
import RNFS from "react-native-fs";
import type {
  PhotoFile
} from "react-native-vision-camera";
import resizeImage from "sharedHelpers/resizeImage";
import { unlink } from "sharedHelpers/util";

const savePhotoToDocumentsDirectory = async (
  cameraPhoto: PhotoFile
) => {
  const path = rotatedOriginalPhotosPath;
  await RNFS.mkdir( path );
  // Move the image with ImageResizer (for legacy reasons, because we
  // used to use it to rotate the photo)
  const image = await resizeImage(
    cameraPhoto.path,
    {
      width: cameraPhoto.width,
      height: cameraPhoto.height,
      outputPath: path
    }
  );
  // Remove original photo
  await unlink( cameraPhoto.path );
  return image;
};

export default savePhotoToDocumentsDirectory;
