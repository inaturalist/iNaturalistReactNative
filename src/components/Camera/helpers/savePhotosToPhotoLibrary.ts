import { CameraRoll } from "@react-native-camera-roll/camera-roll";
import {
  permissionResultFromMultiple,
  READ_WRITE_MEDIA_PERMISSIONS
} from "components/SharedComponents/PermissionGateContainer.tsx";
import { t } from "i18next";
import {
  Alert
} from "react-native";
import { checkMultiple, RESULTS } from "react-native-permissions";
import { log } from "sharedHelpers/logger";

import { displayName as appName } from "../../../../app.json";

const logger = log.extend( "savePhotosToPhotoLibrary" );

// Save URIs to camera photo library (if a photo was taken using the app,
// we want it accessible in the camera's folder, as if the user has taken those photos
// via their own camera app).
// One could argue this is a private method and shouldn't be exported and
// doesn't need to be tested... but hooks are complicated and this hook might
// be too complicated, so this at least makes it easy to test this one part
// ~~~kueda20240614
// $FlowIgnore
export async function savePhotosToPhotoLibrary(
  uris: [string],
  location: Object
) {
  const readWritePermissionResult = permissionResultFromMultiple(
    await checkMultiple( READ_WRITE_MEDIA_PERMISSIONS )
  );
  const savedPhotoUris = await uris.reduce(
    async ( memo, uri ) => {
      const savedUris = await memo;
      try {
        const saveOptions = {};
        // One quirk of CameraRoll is that if you want to write to an album, you
        // need readwrite permission, but we don't want to ask for that here
        // b/c it might come immediately after asking for *add only*
        // permission, so we're checking to see if we have that permission
        // and skipping the album if we don't
        if ( readWritePermissionResult === RESULTS.GRANTED ) {
          saveOptions.type = "photo";
          // Note: we do not translate our brand name, so this should not be
          // globalized
          saveOptions.album = appName;
        }
        if ( location ) {
          saveOptions.latitude = location.latitude;
          saveOptions.longitude = location.longitude;
          saveOptions.horizontalAccuracy = location.positional_accuracy;
        }
        const savedPhotoUri = await CameraRoll.save( uri, saveOptions );
        savedUris.push( savedPhotoUri );
        return savedUris;
      } catch ( cameraRollSaveError ) {
        // should never get here since in usePrepareStoreAndNavigate we check for device full
        // and skip saving to photo library
        if (
          cameraRollSaveError.message.match( /No space left on device/ )
          || cameraRollSaveError.message.match( /PHPhotosErrorDomain error 3305/ )
        ) {
          Alert.alert(
            t( "Not-enough-space-left-on-device" ),
            t( "Not-enough-space-left-on-device-try-again" ),
            [{ text: t( "OK" ) }]
          );
          return savedUris;
        }
        // This means an iOS user denied access
        // (https://developer.apple.com/documentation/photokit/phphotoserror/code/accessuserdenied).
        // In theory we should not even have called this function when that
        // happens, but we're still seeing this in the logs. They should be
        // prompted to grant permission the next time they try so this is
        // probably safe to ignore.
        if ( !cameraRollSaveError.message.match( /error 3311/ ) ) {
          logger.error( cameraRollSaveError );
          return savedUris;
        }
        throw cameraRollSaveError;
      }
    },
    // We need the initial value even if we're not using it, otherwise reduce
    // will treat the first item in the array as the initial value and not
    // call the reducer function on it
    Promise.resolve( [] )
  );
  return savedPhotoUris;
}

export default savePhotosToPhotoLibrary;
