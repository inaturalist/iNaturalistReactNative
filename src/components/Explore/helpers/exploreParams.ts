// Collection of helpers that help prepare API request parameters

import type {
  ApiObservationsSearchParams,
  ApiObservationsSearchResponse
} from "api/types";
import { addSeconds, formatISO, parseISO } from "date-fns";
import { last } from "lodash";

interface ApiObservationsSearchParamsForInfiniteQuery extends ApiObservationsSearchParams {
  pageParam: number | string | undefined | null;
}

// Gets the next page param for Explore given the last page
//
// Pagination using our API is complicated. The page param will only work
// for the first 10k results. Frequently-updated result sets (like the
// default), can potentially show duplicate observations on subsequent
// pages using the page param as well. To avoid both of those issues, we're
// paginating results using the id_below / id_above or d1 / d2 params when
// possible, i.e. when sorting by date created or date observed. Other
// sorts must use the page param and suffer the limitations described
// above.
function getNextPageParamForExplore(
  lastPage: ApiObservationsSearchResponse,
  params: ApiObservationsSearchParams
) {
  const lastObs = last( lastPage.results );
  const orderBy = params.order_by;

  // Returning null tells useInfiniteQuery there is no next page, so the
  // query method should not even run. If there are no observations in the
  // results, we're done and can stop requesting new pages.
  if ( !lastObs ) return null;

  // Datetime sorts need to use d1 / d2 or created_d1 / created_d2 to
  // paginate results, so were storing a datetime from the last observation
  // as the "page" and we'll use that to adjust the query when we perform
  // it
  if ( ["observed_on", "created_at"].includes( String( orderBy ) ) ) {
    const lastObsDate = orderBy === "observed_on"
      ? lastObs?.time_observed_at
      : lastObs?.created_at;

    // If there are results but the last one doesn't have a datetime, we're
    // also done... but this is probably impossible.
    if ( !lastObsDate ) {
      return null;
    }

    const lastObsDateParsed = parseISO( lastObsDate );
    // Adding / subtracting a second helps us not include this last
    // observation on the next page of results
    const newObsDate = addSeconds( lastObsDateParsed, params.order === "asc"
      ? 1
      : -1 );
    return formatISO( newObsDate );
  }

  // Any sort that isn't observed_on or created_at and isn't the default
  // (blank, meaning created_at) should use basic pagination.
  if ( typeof ( orderBy ) === "string" ) {
    return lastPage.page + 1;
  }

  // If we got here that means orderBy is undefined or null, i.e. the
  // default sort order by id
  return lastObs?.id;
}

// Actually modifieds the Explore API params before the request to include the
// page params
function addPageParamsForExplore( params: ApiObservationsSearchParamsForInfiniteQuery ) {
  const { pageParam } = params;
  const newParams = { ...params };

  // For the purposes of Explore, an undefined page or a page of 0 is the
  // default state before any data has loaded, so those values mean we're
  // requesting the first page
  const requestingFirstPage = typeof ( pageParam ) === "undefined" || pageParam === 0;
  if ( requestingFirstPage ) {
    // For the first page and only for the first page, we need to retrieve
    // the georaphic bounds of the results so we can pan and zoom the map
    // to contain them
    newParams.return_bounds = true;
  } else if ( newParams.order_by === "observed_on" && typeof ( pageParam ) === "string" ) {
    // If we're ordering by date observed, we are "paginating" by date and
    // getNextPageParam will have set the pageParam to a date string from
    // the last obs in the previous page
    if ( params.order === "asc" ) {
      newParams.d1 = pageParam;
    } else {
      newParams.d2 = pageParam;
    }
  } else if ( newParams.order_by === "created_at" && typeof ( pageParam ) === "string" ) {
    // If we're ordering by date created, we are "paginating" by date and
    // getNextPageParam will have set the pageParam to a date string from
    // the last obs in the previous page
    if ( params.order === "asc" ) {
      newParams.created_d1 = pageParam;
    } else {
      newParams.created_d2 = pageParam;
    }
  } else if ( typeof ( newParams.order_by ) === "string" && typeof ( pageParam ) === "number" ) {
    // If this is any kind of sort other than the date sorts and isn't the
    // default (blank), assume basic pagination using the page param
    newParams.page = pageParam;
  } else {
    // If we're using default sort order by id, getNextPageParam will have
    // set pageParam to a serial id
    newParams.id_below = Number( pageParam );
  }
  return newParams;
}

export {
  addPageParamsForExplore,
  getNextPageParamForExplore
};
