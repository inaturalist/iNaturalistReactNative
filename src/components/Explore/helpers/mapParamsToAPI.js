// @flow

import {
  ESTABLISHMENT_MEAN,
  MEDIA,
  PHOTO_LICENSE,
  REVIEWED,
  SORT_BY,
  WILD_STATUS
} from "providers/ExploreContext";

const mapParamsToAPI = ( params: Object, currentUser: Object ): Object => {
  const RESEARCH = "research";
  const NEEDS_ID = "needs_id";
  const CASUAL = "casual";

  const CREATED_AT = "created_at"; // = date uploaded at
  const OBSERVED_ON = "observed_on";
  const VOTES = "votes";

  const DESC = "desc";
  const ASC = "asc";

  // Remove all params that are falsy
  const filteredParams = Object.entries( params ).reduce(
    ( newParams, [key, value] ) => {
      if ( value ) {
        newParams[key] = value;
      }
      return newParams;
    },
    {}
  );

  // DATE_UPLOADED_NEWEST is the default sort order
  filteredParams.order_by = CREATED_AT;
  filteredParams.order = DESC;
  if ( params.sortBy === SORT_BY.DATE_UPLOADED_OLDEST ) {
    filteredParams.order_by = CREATED_AT;
    filteredParams.order = ASC;
  }
  if ( params.sortBy === SORT_BY.DATE_OBSERVED_NEWEST ) {
    filteredParams.order_by = OBSERVED_ON;
    filteredParams.order = DESC;
  }
  if ( params.sortBy === SORT_BY.DATE_OBSERVED_OLDEST ) {
    filteredParams.order_by = OBSERVED_ON;
    filteredParams.order = ASC;
  }
  if ( params.sortBy === SORT_BY.MOST_FAVED ) {
    filteredParams.order_by = VOTES;
    filteredParams.order = DESC;
  }

  filteredParams.quality_grade = [];
  if ( params.researchGrade ) {
    filteredParams.quality_grade.push( RESEARCH );
  }
  if ( params.needsID ) {
    filteredParams.quality_grade.push( NEEDS_ID );
  }
  if ( params.casual ) {
    filteredParams.quality_grade.push( CASUAL );
    delete filteredParams.verifiable;
  }

  if ( filteredParams.months ) {
    filteredParams.month = filteredParams.months;
    delete filteredParams.months;
  }

  // MEDIA.ALL is the default media filter and for it we don't need to pass any params
  if ( params.media === MEDIA.PHOTOS ) {
    filteredParams.photos = true;
  } else if ( params.media === MEDIA.SOUNDS ) {
    filteredParams.sounds = true;
  } else if ( params.media === MEDIA.NONE ) {
    filteredParams.photos = false;
    filteredParams.sounds = false;
  }

  // ESTABLISHMENT_MEAN.ANY is the default here and for it we don't need to pass any params
  if ( params.establishmentMean === ESTABLISHMENT_MEAN.NATIVE ) {
    filteredParams.native = true;
  } else if ( params.establishmentMean === ESTABLISHMENT_MEAN.INTRODUCED ) {
    filteredParams.introduced = true;
  } else if ( params.establishmentMean === ESTABLISHMENT_MEAN.ENDEMIC ) {
    filteredParams.endemic = true;
  }

  if ( params.wildStatus === WILD_STATUS.WILD ) {
    filteredParams.captive = false;
  } else if ( params.wildStatus === WILD_STATUS.CAPTIVE ) {
    filteredParams.captive = true;
  }

  if ( params.reviewedFilter === REVIEWED.REVIEWED ) {
    filteredParams.reviewed = true;
    filteredParams.viewer_id = currentUser?.id;
  } else if ( params.reviewedFilter === REVIEWED.UNREVIEWED ) {
    filteredParams.reviewed = false;
    filteredParams.viewer_id = currentUser?.id;
  }

  if ( params.photoLicense !== PHOTO_LICENSE.ALL ) {
    // How license filter maps to the API
    const licenseParams = {
      [PHOTO_LICENSE.CC0]: "cc0",
      [PHOTO_LICENSE.CCBY]: "cc-by",
      [PHOTO_LICENSE.CCBYNC]: "cc-by-nc",
      [PHOTO_LICENSE.CCBYSA]: "cc-by-sa",
      [PHOTO_LICENSE.CCBYND]: "cc-by-nd",
      [PHOTO_LICENSE.CCBYNCSA]: "cc-by-nc-sa",
      [PHOTO_LICENSE.CCBYNCND]: "cc-by-nc-nd"
    };
    filteredParams.photo_license = licenseParams[params.photoLicense];
  }

  delete filteredParams.taxon;
  delete filteredParams.place_guess;
  delete filteredParams.placeMode;
  delete filteredParams.user;
  delete filteredParams.project;
  delete filteredParams.sortBy;
  delete filteredParams.researchGrade;
  delete filteredParams.needsID;
  delete filteredParams.casual;
  delete filteredParams.dateObserved;
  delete filteredParams.dateUploaded;
  delete filteredParams.media;
  delete filteredParams.establishmentMean;
  delete filteredParams.wildStatus;
  delete filteredParams.reviewedFilter;
  delete filteredParams.photoLicense;
  delete filteredParams.place;

  return filteredParams;
};

export default mapParamsToAPI;
