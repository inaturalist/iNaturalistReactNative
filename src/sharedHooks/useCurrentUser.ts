import { RealmContext } from "providers/contexts";
import { useEffect, useState } from "react";
import Realm from "realm";
import type { RealmUser } from "realmModels/types";

const { useRealm } = RealmContext;

const useCurrentUser = ( ): RealmUser | null => {
  const realm = useRealm( );
  const [currentUser, setCurrentUser] = useState<RealmUser | null>( null );
  useEffect( ( ) => {
    const realmResults = realm.objects( "User" ).filtered( "signedIn == true" );

    // Sets the current user if one is detected, unsets it if not
    const listener = (
      collection: [Realm.Object<RealmUser>],
      changes: Realm.CollectionChangeSet
    ) => {
      if (
        ( changes.deletions && changes.deletions.length > 0 )
        || !collection[0]?.isValid( )
      ) {
        setCurrentUser( null );
      } else {
        setCurrentUser( collection[0] );
      }
    };

    // Run the listener on the results for the first time
    listener( realmResults, {} );

    // Add the listener
    realmResults?.addListener( listener );

    // Remove the listener when this dismounts
    return ( ) => {
      realmResults?.removeListener( listener );
    };
  }, [realm] );

  return currentUser?.isValid( )
    ? currentUser
    : null;
};

export default useCurrentUser;
