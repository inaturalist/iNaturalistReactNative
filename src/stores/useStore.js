import _ from "lodash";
import { MMKV } from "react-native-mmkv";
import { create } from "zustand";
import { createJSONStorage, persist } from "zustand/middleware";

import createExploreSlice from "./createExploreSlice";
import createLayoutSlice from "./createLayoutSlice";
import createMyObsSlice from "./createMyObsSlice";
import createObservationFlowSlice from "./createObservationFlowSlice";
import createRootExploreSlice from "./createRootExploreSlice";
import createSyncObservationsSlice from "./createSyncObservationsSlice";
import createUploadObservationsSlice from "./createUploadObservationsSlice";

export const storage = new MMKV( );

// TODO do *not* export this. This allows any consumer to overwrite *any* part
// of state, circumventing any getter/setter logic we have in the stores. If
// you need to modify state, you should be doing so through a store.
export const zustandStorage = {
  setItem: ( name, value ) => storage.set( name, value ),
  getItem: name => {
    const value = storage.getString( name ) || storage.getNumber( name );
    return value ?? null;
  },
  removeItem: name => storage.delete( name )
};

// Using slices to separate store for Explore and Observation creation flow
// https://docs.pmnd.rs/zustand/guides/slices-pattern
const useStore = create( persist(
  ( ...args ) => {
    // Let's make our slices
    const slices = [
      createExploreSlice( ...args ),
      createLayoutSlice( ...args ),
      createMyObsSlice( ...args ),
      createObservationFlowSlice( ...args ),
      createRootExploreSlice( ...args ),
      createSyncObservationsSlice( ...args ),
      createUploadObservationsSlice( ...args )
    ];

    // Now let's make sure they're not clobbering each other because
    // everything in Zustand state exists in a single namespace and
    // clobbering can happen silently... which is bad.
    const allKeys = slices.map( slice => Object.keys( slice ) ).flat( );
    const keyCounts = allKeys.reduce(
      ( memo, curr ) => {
        memo[curr] ||= 0;
        memo[curr] += 1;
        return memo;
      },
      {}
    );
    const nonUniqueKeys = Object.keys( keyCounts ).reduce(
      ( memo, curr ) => {
        if ( keyCounts[curr] > 1 ) memo.push( curr );
        return memo;
      },
      []
    );
    if ( nonUniqueKeys.length > 0 ) {
      throw new Error(
        `You have multiple Zustand slices with the following keys: ${nonUniqueKeys}`
      );
    }

    // All good? Now let's combine them into one enormous hideous object
    return slices.reduce(
      ( memo, curr ) => ( { ...memo, ...curr } ),
      {}
    );
  },
  {
    name: "persisted-zustand",
    partialize: state => ( {
      // Vestigial un-namespaced values in the layout slice
      isAdvancedUser: state.isAdvancedUser,
      obsDetailsTab: state.obsDetailsTab,

      // Dynamically select all values in the layout slice's namespace
      layout: ( Object.keys( state.layout ).reduce( ( memo, key ) => {
        if ( typeof ( state.layout[key] ) !== "function" ) {
          memo[key] = state.layout[key];
        }
        return memo;
      }, {} ) )
    } ),
    storage: createJSONStorage( () => zustandStorage ),
    // We need to deep merge to persist nested objects, like layout
    // https://zustand.docs.pmnd.rs/middlewares/persist#persisting-a-state-with-nested-objects
    merge: ( persisted, current ) => _.merge( current, persisted )
  }
) );

export default useStore;
