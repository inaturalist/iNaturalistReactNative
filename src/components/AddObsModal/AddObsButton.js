// @flow

import { CommonActions, useNavigation } from "@react-navigation/native";
import AddObsModal from "components/AddObsModal/AddObsModal.tsx";
import { Modal } from "components/SharedComponents";
import GradientButton from "components/SharedComponents/Buttons/GradientButton.tsx";
import { t } from "i18next";
import { getCurrentRoute } from "navigation/navigationUtils.ts";
import * as React from "react";
import { log } from "sharedHelpers/logger";
import { useLayoutPrefs } from "sharedHooks";
import useStore from "stores/useStore";

const logger = log.extend( "AddObsButton" );

const AddObsButton = (): React.Node => {
  const [showModal, setModal] = React.useState( false );

  const openModal = React.useCallback( () => setModal( true ), [] );
  const closeModal = React.useCallback( () => setModal( false ), [] );

  const resetObservationFlowSlice = useStore( state => state.resetObservationFlowSlice );
  const { isAllAddObsOptionsMode } = useLayoutPrefs( );
  const navigation = useNavigation( );
  React.useEffect( ( ) => {
    // don't remove this logger.info statement: it's used for internal
    // metrics. isAdvancedUser name is vestigial, changing it will make it
    // impossible to compare with older log data
    logger.info( `isAdvancedUser: ${isAllAddObsOptionsMode}` );
  }, [isAllAddObsOptionsMode] );

  const navAndCloseModal = ( screen, params ) => {
    const currentRoute = getCurrentRoute();
    if ( screen !== "ObsEdit" ) {
      resetObservationFlowSlice( );
    }

    // we need to reset the navigation stack whenever a user navigates from the AddObs wheel,
    // otherwise the user can end up closing out to a previous place in the stack, #1857
    navigation.dispatch(
      CommonActions.reset( {
        index: 0,
        routes: [
          {
            name: "NoBottomTabStackNavigator",
            state: {
              index: 0,
              routes: [
                {
                  name: screen,
                  params: { ...params, previousScreen: currentRoute }
                }
              ]
            }
          }
        ]
      } )
    );

    closeModal( );
  };
  const navToARCamera = ( ) => { navAndCloseModal( "Camera", { camera: "AI" } ); };

  const addObsModal = <AddObsModal closeModal={closeModal} navAndCloseModal={navAndCloseModal} />;

  return (
    <>
      {/* match the animation timing on FadeInView.tsx */}
      <Modal
        animationIn="fadeIn"
        animationOut="fadeOut"
        animationInTiming={250}
        animationOutTiming={250}
        showModal={showModal}
        closeModal={closeModal}
        modal={addObsModal}
      />
      <GradientButton
        sizeClassName="w-[69px] h-[69px] mb-[5px]"
        onPress={isAllAddObsOptionsMode
          ? openModal
          : navToARCamera}
        accessibilityLabel={t( "Add-observations" )}
        accessibilityHint={isAllAddObsOptionsMode
          ? t( "Shows-observation-creation-options" )
          : t( "Opens-AI-camera" )}
        iconName={isAllAddObsOptionsMode && "plus"}
        iconSize={isAllAddObsOptionsMode && 31}
      />
    </>
  );
};

export default AddObsButton;
