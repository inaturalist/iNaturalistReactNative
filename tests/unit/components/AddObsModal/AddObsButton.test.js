import { screen } from "@testing-library/react-native";
import AddObsButton from "components/AddObsModal/AddObsButton";
import React from "react";
import * as useCurrentUser from "sharedHooks/useCurrentUser.ts";
import { zustandStorage } from "stores/useStore";
import factory from "tests/factory";
import { renderComponent } from "tests/helpers/render";
import setStoreStateLayout from "tests/helpers/setStoreStateLayout";

// Mock getCurrentRoute to return ObsList
jest.mock( "navigation/navigationUtils", () => ( {
  getCurrentRoute: () => ( {
    name: "ObsList"
  } )
} ) );

const mockUser = factory( "LocalUser" );

jest.mock( "sharedHooks/useCurrentUser", () => ( {
  __esModule: true,
  default: () => undefined
} ) );

describe( "AddObsButton", () => {
  it( "should not have accessibility errors", () => {
    // const addObsButton = <AddObsButton />;

    // Disabled during the update to RN 0.78
    // expect( addObsButton ).toBeAccessible();
  } );

  it( "renders correctly", () => {
    renderComponent( <AddObsButton /> );
    // Snapshot test
    expect( screen ).toMatchSnapshot();
  } );

  it( "does not render tooltip in default state", () => {
    renderComponent( <AddObsButton /> );

    const tooltipText = screen.queryByText(
      "Press and hold to view more options"
    );
    expect( tooltipText ).toBeFalsy();
  } );
} );

describe( "shows tooltip", () => {
  it( "to logged out users with 2 observations", async () => {
    zustandStorage.setItem( "numOfUserObservations", 2 );

    renderComponent( <AddObsButton /> );

    const tooltipText = await screen.findByText(
      "Press and hold to view more options"
    );
    expect( tooltipText ).toBeTruthy();
  } );

  it( "to logged in users with less than 50 observations", async () => {
    zustandStorage.setItem( "numOfUserObservations", 2 );
    jest.spyOn( useCurrentUser, "default" ).mockImplementation( () => mockUser );

    renderComponent( <AddObsButton /> );

    // Temporarily disabled the tooltip for new users, as it is freezing the app in some cases.
    // const tooltipText = await screen.findByText(
    //   "Press and hold to view more options"
    // );
    // expect( tooltipText ).toBeTruthy();
  } );

  it( "to new users only after they dismissed the account creation card", async () => {
    zustandStorage.setItem( "numOfUserObservations", 1 );
    setStoreStateLayout( {
      justFinishedSignup: true
    } );

    renderComponent( <AddObsButton /> );

    const tooltipText = screen.queryByText(
      "Press and hold to view more options"
    );
    expect( tooltipText ).toBeFalsy();

    setStoreStateLayout( {
      shownOnce: {
        "account-creation": true
      }
    } );

    // Temporarily disabled the tooltip for new users, as it is freezing the app in some cases.
    // const tooltipTextAfter = await screen.findByText(
    //   "Press and hold to view more options"
    // );
    // expect( tooltipTextAfter ).toBeTruthy();
  } );

  it( "to logged in users with more than 50 observations after card dismissal", async () => {
    zustandStorage.setItem( "numOfUserObservations", 51 );

    renderComponent( <AddObsButton /> );

    const tooltipText = screen.queryByText(
      "Press and hold to view more options"
    );
    expect( tooltipText ).toBeFalsy();

    setStoreStateLayout( {
      shownOnce: {
        "fifty-observation": true
      }
    } );

    // Temporarily disabled the tooltip for new users, as it is freezing the app in some cases.
    // const tooltipTextAfter = await screen.findByText(
    //   "Press and hold to view more options"
    // );
    // expect( tooltipTextAfter ).toBeTruthy();
  } );
} );
