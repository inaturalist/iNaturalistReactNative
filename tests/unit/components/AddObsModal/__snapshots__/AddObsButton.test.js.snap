// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddObsButton renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    className="flex-1"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityHint="Opens AI camera."
      accessibilityLabel="Add observations"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          [
            {
              "width": 69,
            },
            {
              "height": 69,
            },
            {
              "marginBottom": 5,
            },
            {
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
            },
            {
              "overflow": "hidden",
            },
            {
              "elevation": 5,
              "shadowColor": "rgba( 0, 0, 0, 0.25 )",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 2,
            },
          ],
        ]
      }
      testID="add-obs-button"
    >
      <BVLinearGradient
        angle={156.95}
        colors={
          [
            4285836288,
            4280909703,
          ]
        }
        endPoint={
          {
            "x": 0.5,
            "y": 1,
          }
        }
        locations={null}
        startPoint={
          {
            "x": 0.5,
            "y": 0,
          }
        }
        useAngle={true}
      >
        <View
          style={
            [
              [
                {
                  "flexGrow": 1,
                },
                {
                  "aspectRatio": 1,
                },
                {
                  "display": "flex",
                },
                {
                  "alignItems": "center",
                },
                {
                  "justifyContent": "center",
                },
              ],
            ]
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#ffffff",
                  "fontSize": 47,
                },
                null,
                {
                  "fontFamily": "INatIcon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            ï…¸
          </Text>
        </View>
      </BVLinearGradient>
    </View>
  </View>
</View>
`;
